### ISSUE TITLE
Investigate LS->beats read timeout

### ISSUE THREAD

**Issue by urso** [logstash-plugins/logstash-input-beats#185]
For all general issues, please provide the following details for fast resolution:

- Version: beats and logstash 5.2

We've got a number of reports about beats->Logstash connection loss due to timeout:

```
ERR Failed to publish events caused by: read tcp : i/o timeout
```

This timeout occurs when beats is waiting for ACK from the input plugin. In case logstash is busy, a heartbeat signal `ACK 0` is supposed to be send every 5 seconds, to signal beats the batch of events being send are still in progress. Plus beats employ a read timeout of 30 seconds. That is, we missed about 6 heartbeat signals (the timeout in beats is configurable).

We should make the `heartbeat` interval configurable and have some integration tests with beats or [go-lumber](https://github.com/elastic/go-lumber) (go-lumber allows to wait for partial ACKs, so we can verify heartbeat ACKs being received).

Interestingly, in all cases users did report LS being busy. e.g. in one incident user did report logstash having very high resource using trying to send events to failing ES cluster over some long period of time. [This one](https://discuss.elastic.co/t/filebeat-duplicate-events/73693/3) has some 'grok timeout' error messages.

Will add more reports coming in (or old ones if I find them) to this ticket.

**Comment by urso** [comment]
Another one: https://discuss.elastic.co/t/metricbeats-timeout/73890

**Comment by ph** [comment]
This is revealling to be an easy fix, the `BeatsHandler` is marked as shareable but has states in it.

**Comment by ph** [comment]
I think #191 also address this problem, will do a bit more testing to make sure it does.

**Comment by ph** [comment]
fixed in #191
